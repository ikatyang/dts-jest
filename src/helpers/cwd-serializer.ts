export const cwd_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(new RegExp(process.cwd(), 'g'), '<cwd>')),
  test: (value: any) =>
    typeof value === 'string' && value.indexOf(process.cwd()) !== -1,
};
