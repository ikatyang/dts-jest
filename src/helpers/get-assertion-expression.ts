import {
  runtime_namespace,
  IRuntimeReportParameters,
  IRuntimeSnapshotParameters,
  TriggerKind,
} from '../definitions';

export const get_assertion_expression = (kind: TriggerKind, line: number) => {
  if (kind === TriggerKind.Show) {
    const report_parameters: IRuntimeReportParameters = {line};
    const report_value = `${runtime_namespace}.report(${JSON.stringify(report_parameters)})`;
    return `console.log(${report_value})`;
  }
  const snapshot_parameters: IRuntimeSnapshotParameters = {line};
  const snapshot_value = `${runtime_namespace}.snapshot(${JSON.stringify(snapshot_parameters)})`;
  return `expect(${snapshot_value}).toMatchSnapshot()`;
};
