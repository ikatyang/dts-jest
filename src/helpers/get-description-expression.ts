import {
  runtime_namespace,
  IRuntimeDescriptionParameters,
} from '../definitions';

export const get_description_expression = (line: number) => {
  const parameters: IRuntimeDescriptionParameters = {
    line,
  };
  return `${runtime_namespace}.description(${JSON.stringify(parameters)})`;
};
