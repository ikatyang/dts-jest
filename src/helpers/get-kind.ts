import {TriggerKind} from '../definitions';

export const get_kind = (flag?: string) => {
  switch (flag) {
    case ':show':
      return TriggerKind.Show;
    case ':skip':
      return TriggerKind.Skip;
    case ':only':
      return TriggerKind.Only;
    case undefined:
      return TriggerKind.None;
    default:
      throw new Error(`Unexpected flag ${JSON.stringify(flag)}`);
  }
};
