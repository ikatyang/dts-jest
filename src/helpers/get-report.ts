import {indent} from '../utils/indent';

// tslint:disable:no-magic-numbers

export const get_report = (reporter_template: string, expression: string, snapshot: string) =>
  reporter_template
    .split(/\{\{(expression|snapshot),([0-9]+)\}\}/g)
    .map((chunk, index, chunks) =>
      (index % 3 === 0)
          ? chunk
          : (index % 3 === 1)
            ? ''
            : (chunks[index - 1] === 'snapshot')
              ? indent(snapshot, +chunk)
              : indent(expression, +chunk),
    )
    .join('');
