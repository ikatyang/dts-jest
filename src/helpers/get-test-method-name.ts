import {TriggerKind} from '../definitions';

export const get_test_method_name = (kind: TriggerKind) => {
  switch (kind) {
    case TriggerKind.None:
    case TriggerKind.Show:
      return 'test';
    case TriggerKind.Only:
      return 'test.only';
    case TriggerKind.Skip:
      return 'test.skip';
    default:
      throw new Error(`Unexpected kind ${JSON.stringify(kind)}`);
  }
};
