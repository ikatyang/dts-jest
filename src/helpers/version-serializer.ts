const version_regex = /\bv[0-9]+\.[0-9]+\.[0-9]+\b/g;

export const version_serializer: jest.SnapshotSerializerPlugin = {
  print: (value: string, serializer) =>
    serializer(value.replace(version_regex, 'vX.Y.Z-mocked')),
  test: (value: any) => typeof value === 'string' && version_regex.test(value),
};
