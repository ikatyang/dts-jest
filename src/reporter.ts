import * as path from 'path';
import { config_namespace, JestConfig } from './definitions';
import { get_config } from './utils/get-config';

export class Reporter implements jest.Reporter {
  // tslint:disable-next-line:naming-convention prefer-function-over-method
  public onRunComplete(contexts: Set<jest.Context>) {
    process.stdout.write('\n');

    contexts.forEach(context => {
      const {
        typescript: { version: ts_version },
        typescript_path: ts_path,
      } = get_config(
        (context.config as JestConfig).globals[config_namespace],
        context.config.rootDir,
      );

      // istanbul ignore next
      const context_name =
        contexts.size > 1 ? ` ${path.basename(context.config.rootDir)}` : '';

      process.stdout.write(
        `[dts-jest]${context_name} using TypeScript v${ts_version} from ${ts_path}\n`,
      );
    });

    process.stdout.write('\n');
  }
}
