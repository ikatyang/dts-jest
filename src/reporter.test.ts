import { cwd_serializer } from './helpers/cwd-serializer';
import { version_serializer } from './helpers/version-serializer';
import { Reporter } from './reporter';

expect.addSnapshotSerializer(cwd_serializer);
expect.addSnapshotSerializer(version_serializer);

// tslint:disable-next-line:no-unbound-method
const original_stdout_write = process.stdout.write;
let mocked_stdout_write: jest.MockInstance<any>;

beforeEach(() => {
  mocked_stdout_write = process.stdout.write = jest.fn();
});

afterAll(() => {
  process.stdout.write = original_stdout_write;
});

it('should report correctly', () => {
  const reporter = new Reporter(create_config());
  reporter.onRunComplete(new Set([create_context()]));
  expect(get_write_content()).toMatchSnapshot();
});

function create_context(): jest.Context {
  return {
    config: {
      rootDir: process.cwd(),
      globals: {},
    },
  } as any;
}

function create_config(): jest.GlobalConfig {
  return { useStderr: false } as any;
}

function get_write_content() {
  return mocked_stdout_write.mock.calls.map(x => x.join('')).join('');
}
