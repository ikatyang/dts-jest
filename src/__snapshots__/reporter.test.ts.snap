// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should report correctly 1`] = `
"
[dts-jest] using TypeScript vX.Y.Z-mocked from <cwd>/node_modules/typescript/lib/typescript.js

"
`;
